File: cpp/a12.cpp
#include <iostream>

using namespace std;

long n, k;
long a[100009];

bool check(long m) {
    long sum = 0;
    for (long i = 1; i <= n; i++)
    {
        sum += m / a[i];
    }
    return sum >= k;
}

int main() {

    cin >> n >> k;
    for (long i = 1; i <= n; i++)
    {
        cin >> a[i];
    }

    long l = 1;
    long r = 1000000000;
    while(l < r) {
        long m = (l + r) / 2;
        if (check(m)) r = m;
        else l = m + 1;
    }
    cout << l;
    return 0;
}

File: scala/a12.scala
// import scala.io.StdIn._
import java.util.Scanner

object Main extends App {
  val sc = new Scanner(System.in)
  val n = sc.nextInt()
  val k = sc.nextInt()
  val a = 0 +: Vector.fill(n){sc.nextInt()}

  def search(k: Long): Long = {
    @annotation.tailrec
    def go(k: Long, l: Long, r: Long): Long = {
      val m = (l + r) / 2
      if (l == r) l
      else if (check(m)) go(k, l, m)
      else go(k, m+1, r)
    }
    go(k, 1, 1000000000)
  }

  def check(m: Long): Boolean = {
    val printer = (1 to n).map(i => m/a(i))
    val sum = printer.sum
    sum >= k
  }

  println(search(k))
}
